erpnext/buying/doctype/supplier_scorecard/supplier_scorecard.py

```diff
@@ -0,0 +1,237 @@
+		score = 0
+		for c1 in self.standings:
+			for c2 in self.standings:
+				if c2.min_grade == score:
+					score = c2.max_grade
+		if score < 100 :
+			throw(_('Unable to find score starting at ' + str(score) + '. You need to have standing scores covering 0 to 100'))
+
+	def validate_criteria_weights(self):
+
+		weight = 0
+		for c in self.criteria:
+			weight += c.weight
+
+		if weight != 100:
+			throw(_('Criteria weights must add up to 100%'))
+
+	def calculate_total_score(self):
+		scorecards = frappe.db.sql("""
+			SELECT
+				scp.name
+			FROM
+				`tabSupplier Scorecard Period` scp
+			WHERE
+				scp.scorecard = %(sc)s
+			ORDER BY
+				scp.end_date DESC""",
+				{"sc": self.name}, as_dict=1)
+
+		period = 0
+		total_score = 0
+		total_max_score = 0
+		for scp in scorecards:
+			my_sc = frappe.get_doc('Supplier Scorecard Period', scp.name)
+			my_scp_weight = self.weighting_function
+			my_scp_weight = my_scp_weight.replace('{period_number}', str(period))
+
+			my_scp_maxweight = my_scp_weight.replace('{total_score}', '100')
+			my_scp_weight = my_scp_weight.replace('{total_score}', str(my_sc.total_score))
+
+			max_score = my_sc.calculate_weighted_score(my_scp_maxweight)
+			score = my_sc.calculate_weighted_score(my_scp_weight)
+
+			total_score += score
+			total_max_score += max_score
+			period += 1
+		if total_max_score > 0:
+			self.supplier_score = round(100.0 * (total_score / total_max_score) ,1)
+		else:
+			self.supplier_score =  100
+
+	def update_standing(self):
+		# Get the setup document
+
+		for standing in self.standings:
+			if (not standing.min_grade or (standing.min_grade <= self.supplier_score)) and \
+				(not standing.max_grade or (standing.max_grade > self.supplier_score)):
+				self.status = standing.standing_name
+				self.indicator_color = standing.standing_color
+				self.warn_pos = standing.warn_pos
+				self.warn_rfqs = standing.warn_rfqs
+				self.prevent_rfqs = standing.prevent_rfqs
+				self.prevent_pos = standing.prevent_pos
+				self.notify_supplier = standing.notify_supplier
+				self.notify_employee = standing.notify_employee
+				self.employee_link = standing.employee_link
+
+				#Update supplier standing info
+				frappe.db.set_value("Supplier", self.supplier, "prevent_pos", self.prevent_pos)
+				frappe.db.set_value("Supplier", self.supplier, "prevent_rfqs", self.prevent_rfqs)
+				frappe.db.set_value("Supplier", self.supplier, "warn_rfqs", self.warn_rfqs)
+				frappe.db.set_value("Supplier", self.supplier, "warn_pos", self.warn_pos)
+
+@frappe.whitelist()
+def timeline_data(doctype, name):
+	# Get a list of all the associated scorecards
+	scs = frappe.get_doc(doctype, name)
+	out = {}
+	timeline_data = {}
+	scorecards = frappe.db.sql("""
+		SELECT
+			sc.name
+		FROM
+			`tabSupplier Scorecard Period` sc
+		WHERE
+			sc.scorecard = %(scs)s""",
+			{"scs": scs.name}, as_dict=1)
+
+	for sc in scorecards:
+		sc = frappe.get_doc('Supplier Scorecard Period', sc.name)
```

Comment:
To fetch 2/3 fields value, try to avoid `get_doc`. Use `get_value` instead.

`start_date, end_date = frappe.db.get_value('Supplier Scorecard Period', sc.name, ['start_date', 'end_date'])`